liquidctl/driver/aquacomputer.py

```diff
@@ -0,0 +1,262 @@
+--------------------------------------
+The pump sends a status HID report every second with no initialization
+being required.
+
+The status HID report exposes sensor values such as liquid temperature and
+two groups of fan sensors, for the pump and the optionally connected fan.
+These groups provide RPM speed, voltage, current and power readings. The
+pump additionally exposes +5V and +12V voltage rail readings.
+
+Driver
+------
+Linux has the aquacomputer_d5next driver available since v5.15. Subsequent
+releases have more functionality and support a wider range of devices. If
+present, it's used instead of reading the status reports directly.
+
+Copyright (C) 2022 - Aleksa Savic
+
+SPDX-License-Identifier: GPL-3.0-or-later
+"""
+
+# uses the psf/black style
+
+import logging
+
+from liquidctl.driver.usb import UsbHidDriver
+from liquidctl.error import NotSupportedByDriver
+
+_LOGGER = logging.getLogger(__name__)
+
+_AQC_TEMP_SENSOR_DISCONNECTED = 0x7FFF
+_AQC_FAN_VOLTAGE_OFFSET = 0x02
+_AQC_FAN_CURRENT_OFFSET = 0x04
+_AQC_FAN_POWER_OFFSET = 0x06
+_AQC_FAN_SPEED_OFFSET = 0x08
+
+_AQC_STATUS_READ_ENDPOINT = 0x01
+
+
+def get_unaligned_be16(msg, offset):
+    return msg[offset] << 8 | msg[offset + 1]
+
+
+class Aquacomputer(UsbHidDriver):
+    # Support for hwmon: aquacomputer_d5next, sensors - 5.15+
+
+    DEVICE_D5NEXT = "D5 Next"
+
+    DEVICE_INFO = {
+        DEVICE_D5NEXT: {
+            "type": DEVICE_D5NEXT,
+            "fan_sensors": [0x6C, 0x5F],
+            "temp_sensors": [0x57],
+            "plus_5v_voltage": 0x39,
+            "plus_12v_voltage": 0x37,
+            "temp_sensors_label": ["Liquid temperature"],
+            "fan_speed_label": ["Pump speed", "Fan speed"],
+            "fan_power_label": ["Pump power", "Fan power"],
+            "fan_voltage_label": ["Pump voltage", "Fan voltage"],
+            "fan_current_label": ["Pump current", "Fan current"],
+            "status_report_length": 0x9E,
+        }
+    }
+
+    SUPPORTED_DEVICES = [
+        (0x0C70, 0xF00E, None, "Aquacomputer D5 Next", {"device_info": DEVICE_INFO[DEVICE_D5NEXT]}),
+    ]
+
+    def __init__(self, device, description, device_info, **kwargs):
+        super().__init__(device, description)
+
+        # Read when necessary
+        self._firmware_version = None
+
+        self._device_info = device_info
+
+    def initialize(self, **kwargs):
+        """Initialize the device and the driver.
+
+        This method should be called every time the system boots, resumes from
+        a suspended state, or if the device has just been (re)connected.  In
+        those scenarios, no other method, except `connect()` or `disconnect()`,
+        should be called until the device and driver has been (re-)initialized.
+
+        Returns None or a list of `(property, value, unit)` tuples, similarly
+        to `get_status()`.
+        """
+
+        fw = self.firmware_version
+        _LOGGER.debug("raw firmware version: %r", fw)
```

Comment:
If the raw firmware version is identical to our human-readable output, it isn't necessary to log it.